numpy/f2py/tests/test_crackfortran.py

```diff
@@ -276,3 +277,49 @@ class TestUnicodeComment(util.F2PyTest):
     )
     def test_encoding_comment(self):
         self.module.foo(3)
+
+class TestNameArgsPatternBacktracking:
+    @pytest.mark.parametrize(
+        ['adversary'],
+        [
+            ('@)@bind@(@',),
+            ('@)@bind                         @(@',),
+            ('@)@bind foo bar baz@(@',)
+        ]
+    )
+    def test_nameargspattern_backtracking(self, adversary):
+        '''address ReDOS vulnerability:
+        https://github.com/numpy/numpy/issues/23338'''
+        last_median = 0.
+        trials_per_count = 128
+        start_reps, end_reps = 15, 25
+        times_median_doubled = 0
+        for ii in range(start_reps, end_reps):
+            repeated_adversary = adversary * ii
+            times = []
+            for _ in range(trials_per_count):
+                t0 = time.perf_counter()
+                mtch = nameargspattern.search(repeated_adversary)
+                times.append(time.perf_counter() - t0)
+            # We should use a measure of time that's resilient to outliers.
+            # Times jump around a lot due to the CPU's scheduler.
+            median = np.median(times)
+            assert not mtch
+            # if the adversary is capped with @)@, it becomes acceptable
+            # according to the old version of the regex.
+            # that should still be true.
+            good_version_of_adversary = repeated_adversary + '@)@'
+            assert nameargspattern.search(good_version_of_adversary)
+            if ii > start_reps:
+                # the hallmark of exponentially catastrophic backtracking
+                # is that runtime doubles for every added instance of
+                # the problematic pattern.
+                times_median_doubled += median > 2 * last_median
+                # also try to rule out non-exponential but still bad cases
+                # arbitrarily, we should set a hard limit of 10ms as too slow
+                assert median < trials_per_count * 0.01
+            last_median = median
+        # we accept that maybe the median might double once, due to
+        # the CPU scheduler acting weird or whatever. More than that
+        # seems suspicious.
+        assert times_median_doubled < 2
```

Comment:
Saw this test fail at least once recently on CI.  We may want to make it more robust, or maybe also just remove it, at least if that happens more often.